Fix excludeFromCache not working with SiteTreePublishingEngine

nikrolls opened the pull request:
`URLArrayObject#addObject` adds metadata to the query string of the URL segment that's added to the queue. `#excludeFromCache` passes this query string through to `SiteTree::get_by_link`, which doesn't understand query strings and returns nothing.

There is also an issue with Subsites: SiteTreePublishingEngine disables Subsite filtering, meaning the `SiteTree::get_by_link` will always return the first URL that matches rather than the one indicated by the URL's metadata.

This changes fixes both issues by trying to retrieve the object first using `URLArrayObject#getObject`, and then falling back to a clean (query-string-free) URL passed to `SiteTree::get_by_link`.

dhensby commented:
Thanks for this and the tests. Would you mind writing out the indentation, please?

nikrolls commented:
@dhensby, I discovered another issue with this function when it came to Subsites. It's related, so I've updated this PR with an implementation that fixes both issues. I've also fixed the tabs (sorry about that!).

dhensby commented:
thanks
